src/Type/TypeCombinator.php

```diff
@@ -165,106 +162,26 @@ public static function union(Type ...$types): Type
 			if ($types[$i] instanceof StringType && !$types[$i] instanceof ClassStringType) {
 				$hasGenericScalarTypes[ConstantStringType::class] = true;
 			}
-			if ($types[$i] instanceof IntersectionType) {
-				$intermediateArrayType = null;
-				$intermediateAccessoryTypes = [];
-				foreach ($types[$i]->getTypes() as $innerType) {
-					if ($innerType instanceof TemplateType) {
-						continue 2;
-					}
-					if ($innerType instanceof ArrayType) {
-						$intermediateArrayType = $innerType;
-						continue;
-					}
-					if ($innerType instanceof AccessoryType || $innerType instanceof CallableType) {
-						if ($innerType instanceof HasOffsetValueType) {
-							$intermediateAccessoryTypes[sprintf('hasOffsetValue(%s)', $innerType->getOffsetType()->describe(VerbosityLevel::cache()))][] = $innerType;
-							continue;
-						}
-
-						$intermediateAccessoryTypes[$innerType->describe(VerbosityLevel::cache())][] = $innerType;
-						continue;
-					}
-				}
-
-				if ($intermediateArrayType !== null) {
-					$arrayTypes[] = $intermediateArrayType;
-					$arrayAccessoryTypes[] = $intermediateAccessoryTypes;
-					unset($types[$i]);
-					continue;
-				}
-			}
-			if (!$types[$i] instanceof ArrayType) {
+			if (!$types[$i]->isArray()->yes()) {
 				continue;
 			}
 
 			$arrayTypes[] = $types[$i];
-
-			$intermediateAccessoryTypes = [];
-			if ($types[$i]->isIterableAtLeastOnce()->yes()) {
-				$nonEmpty = new NonEmptyArrayType();
-				$intermediateAccessoryTypes[$nonEmpty->describe(VerbosityLevel::cache())] = [$nonEmpty];
-			}
-			if ($types[$i]->isList()->yes() && AccessoryArrayListType::isListTypeEnabled()) {
-				$list = new AccessoryArrayListType();
-				$intermediateAccessoryTypes[$list->describe(VerbosityLevel::cache())] = [$list];
-			}
-			$arrayAccessoryTypes[] = $intermediateAccessoryTypes;
 			unset($types[$i]);
 		}
 
 		foreach ($scalarTypes as $classType => $scalarTypeItems) {
 			$scalarTypes[$classType] = array_values($scalarTypeItems);
 		}
 
-		/** @var ArrayType[] $arrayTypes */
-		$arrayTypes = $arrayTypes;
-
-		$commonArrayAccessoryTypesKeys = [];
-		if (count($arrayAccessoryTypes) > 1) {
-			$commonArrayAccessoryTypesKeys = array_keys(array_intersect_key(...$arrayAccessoryTypes));
-		} elseif (count($arrayAccessoryTypes) > 0) {
-			$commonArrayAccessoryTypesKeys = array_keys($arrayAccessoryTypes[0]);
-		}
-
-		$arrayAccessoryTypesToProcess = [];
-		foreach ($commonArrayAccessoryTypesKeys as $commonKey) {
-			$typesToUnion = [];
-			foreach ($arrayAccessoryTypes as $array) {
-				foreach ($array[$commonKey] as $arrayAccessoryType) {
-					$typesToUnion[] = $arrayAccessoryType;
-				}
-			}
-			$arrayAccessoryTypesToProcess[] = self::union(...$typesToUnion);
-		}
```

Comment:
this logic is replaced by `count($arrayType) === count($accessoryType[$key])` in
https://github.com/phpstan/phpstan-src/pull/2118/files#diff-fef252fe8071200a31b3d282d1442135d62297098c9b20fd731f3888398ee973R549-R562

also, there is no need to union accessoryTypes except `HasOffsetValueType`, because the type are all the same.